tests/unittests/Foundation/NSDecimalNumberTests.mm

```diff
@@ -0,0 +1,356 @@
+}
+
+TEST(NSDecimalNumber, NSDecimalCompare) {
+    NSDecimalNumber* dNumber = [[[NSDecimalNumber alloc] initWithUnsignedLongLong:9223372534876] autorelease];
+    NSDecimal right = [dNumber decimalValue];
+
+    NSDecimal left = {};
+    unsigned long long value = 10;
+    memcpy(&left._mantissa, &value, NSDecimalMaxSize * sizeof(unsigned short));
+    left._exponent = 4;
+    left._length = 1;
+    EXPECT_EQ(NSOrderedAscending, NSDecimalCompare(&left, &right));
+
+    //
+    right = [[[[NSDecimalNumber alloc] initWithUnsignedLongLong:9223372534876] autorelease] decimalValue];
+    NSDecimal right2 = [[NSDecimalNumber numberWithUnsignedLongLong:9223372534876] decimalValue];
+
+    left = [[[[NSDecimalNumber alloc] initWithInteger:-1] autorelease] decimalValue];
+
+    EXPECT_EQ(NSOrderedAscending, NSDecimalCompare(&left, &right));
+    EXPECT_EQ(NSOrderedSame, NSDecimalCompare(&right, &right));
+    EXPECT_EQ(NSOrderedSame, NSDecimalCompare(&right, &right2));
+    EXPECT_EQ(NSOrderedSame, NSDecimalCompare(&left, &left));
+
+    //
+    right = [[[[NSDecimalNumber alloc] initWithInteger:-50] autorelease] decimalValue];
+
+    left = [[[[NSDecimalNumber alloc] initWithInteger:-1] autorelease] decimalValue];
+
+    EXPECT_EQ(NSOrderedDescending, NSDecimalCompare(&left, &right));
+
+    //
+    right = [[[[NSDecimalNumber alloc] initWithInteger:-6595] autorelease] decimalValue];
+
+    left = [[[[NSDecimalNumber alloc] initWithInteger:-6596] autorelease] decimalValue];
+
+    EXPECT_EQ(NSOrderedAscending, NSDecimalCompare(&left, &right));
+
+    //
+
+    left = [[[[NSDecimalNumber alloc] initWithInteger:-5] autorelease] decimalValue];
+    right = [[[[NSDecimalNumber alloc] initWithLongLong:3] autorelease] decimalValue];
+    EXPECT_EQ(NSOrderedAscending, NSDecimalCompare(&left, &right));
+    EXPECT_EQ(NSOrderedDescending, NSDecimalCompare(&right, &left));
+
+    left = [[[[NSDecimalNumber alloc] initWithDouble:(2 * DBL_MAX)] autorelease] decimalValue];
+    right = [[[[NSDecimalNumber alloc] initWithDouble:5] autorelease] decimalValue];
+    EXPECT_EQ(NSOrderedDescending, NSDecimalCompare(&left, &right));
+}
+
+TEST(NSDecimalNumber, NSNumberCompare) {
+    EXPECT_OBJCEQ([NSNumber numberWithDouble:55.0], [NSDecimalNumber numberWithDouble:55.0]);
+
+    //
+    EXPECT_OBJCEQ([[[NSDecimalNumber alloc] initWithMantissa:0 exponent:0 isNegative:YES] autorelease],
+                  [[[NSDecimalNumber alloc] initWithMantissa:0 exponent:0 isNegative:YES] autorelease]);
+
+    //
+    EXPECT_EQ(NSOrderedAscending,
+              [[[[NSDecimalNumber alloc] initWithMantissa:0 exponent:0 isNegative:YES] autorelease]
+                  compare:[NSNumber numberWithDouble:55.0]]);
+
+    //
+    EXPECT_EQ(NSOrderedDescending,
+              [[NSNumber numberWithDouble:55.0]
+                  compare:[[[NSDecimalNumber alloc] initWithMantissa:0 exponent:0 isNegative:YES] autorelease]]);
+}
```

Comment:
#1062

---
In reply to: [81399278](https://github.com/Microsoft/WinObjC/pull/1032#discussion_r81399278) [](ancestors = 81399278,81398501,81398304,81398086)